nerfstudio/model_components/renderers.py

```diff
@@ -332,3 +332,32 @@ def forward(
         """Calculate normals along the ray."""
         n = torch.sum(weights * normals, dim=-2)
         return n
+
+
+class LambertianShadingRenderer(nn.Module):
+    """Calculate Lambertian shading."""
+
+    @classmethod
+    def forward(
+        cls,
+        rgb: TensorType["bs":..., "num_samples", 3],
+        normals: TensorType["bs":..., "num_samples", 3],
+        light_direction: TensorType["bs", 3],
+        ratio: float = 1.0,
+    ):
+        """Calculate Lambertian shading.
+
+        Args:
+            rgb: Accumulated rgb along a ray.
+            normals: Accumulated normals along a ray.
+            light_direction: Direction of light source.
+            ratio: ambient lighting (1.0) vs. Lambertian shading (0.0) ratio
+
+        Returns:
+            Textureless Lambertian shading, Lambertian shading
+        """
+        lambertian = ratio + (1 - ratio) * (normals.detach() @ light_direction).clamp(min=0)
```

Comment:
Do we always want to detach normals? Maybe we should add an argument that allows the user to choose.

Reply:
I didn’t see this earlier, but the reason for detaching the normals was that they get supervised by your SDS loss in Dreamfusion if you don’t detach them. The optimization then makes all your normals the same direction which basically turns off lambertian shading. Maybe it’s best to just turn on “detached normals” or something for the Dreamfusion implementation

Reply:
do you remember if you guys ever just tried using the normals network that stable dreamfusion used? I think they were using the sds loss to supervise normals.